Sources/OpenCombine/Publishers/Publishers.DropWhile.swift

```diff
@@ -46,88 +45,107 @@ extension Publishers {
+    typealias GenericPredicate = (Input) -> Result<Bool, Downstream.Failure>
 
-    private var _downstream: Downstream
-    private let _predicate: (Input) throws -> Bool
-    private var _predicateReturnedFalse = false
-    private var _upstreamSubscription: Subscription?
-    private var _demand: Subscribers.Demand = .none
-
-    var description: String { return "DropWhile" }
-
-    var customMirror: Mirror { return Mirror(self, children: EmptyCollection()) }
+    var genericPredicate: GenericPredicate?
+    var demand: Subscribers.Demand = .none
 
-    init(downstream: Downstream, predicate: @escaping (Input) throws -> Bool) {
-        _downstream = downstream
-        _predicate = predicate
+    init(downstream: Downstream, genericPredicate: @escaping GenericPredicate) {
+        self.genericPredicate = genericPredicate
+        super.init(downstream: downstream)
     }
 
+    var description: String { return "DropWhile" }
+
     func receive(subscription: Subscription) {
-        _upstreamSubscription = subscription
+        upstreamSubscription = subscription
         subscription.request(.max(1))
-        _downstream.receive(subscription: self)
+        downstream.receive(subscription: self)
     }
 
     func receive(_ input: Input) -> Subscribers.Demand {
 
-        if _predicateReturnedFalse {
-            return _downstream.receive(input)
+        guard let test = self.genericPredicate else {
+            return downstream.receive(input)
         }
 
-        do {
-            if try !_predicate(input) {
-                _predicateReturnedFalse = true
-                return _demand + _downstream.receive(input) - 1
-            } else {
-                return .max(1)
-            }
-        } catch {
-            // Safe to force unwrap here — predicate throws only if we're within
-            // a TryDropWhile, and its (and its downstream's) Failure type is always
-            // plain Error
-            _downstream.receive(completion: .failure(error as! Downstream.Failure))
+        switch test(input) {
+        case .success(true):
+            return .max(1)
+        case .success(false):
+            self.genericPredicate = nil
+            return demand + downstream.receive(input) - 1
+        case .failure(let error):
+            downstream.receive(completion: .failure(error))
             cancel()
             return .none
         }
     }
 
-    func receive(completion: Subscribers.Completion<Failure>) {
-        switch completion {
-        case .finished:
-            _downstream.receive(completion: .finished)
-        case .failure(let error):
-            // Safe to force unwrap here, since Downstream.Failure can be
-            // either Upstream.Failure or Error
-            _downstream.receive(completion: .failure(error as! Downstream.Failure))
+    func request(_ demand: Subscribers.Demand) {
+        if genericPredicate == nil {
+            upstreamSubscription?.request(demand)
```

Comment:
It's a little unclear to me what's happening here and why.  It could be I don't understand how demand works totally yet.  Is it worth leaving a comment explaining why you don't pass demand on to the upstream when the predicate is `nil`?

Reply:
Yeah, it's not obvious at first glance. Will add a comment.